### PR TITLE
[MRG] Allow base_estimators to be strings

### PR DESCRIPTION
**Pull request by MechCoder**
1. Allow base_estimator to be one of "gp", "et", "rf" or "gbrt", by default "gp"
2. Add a function `cook_skopt_estimator` in utils that allows to cook the default skopt estimators.

### CODE REVIEW

**Comment by MechCoder** [comment]
Fixes https://github.com/scikit-optimize/scikit-optimize/issues/390 and https://github.com/scikit-optimize/scikit-optimize/issues/338

**Comment by iaroslav-ai** [comment]
This is nice

**Comment by iaroslav-ai** [comment]
Super sorry for the closing reopening! I accidentally pressed the button :(

**Comment by MechCoder** [comment]
Why do the tests take so much time to run :-|

File: skopt/optimizer/optimizer.py [hunk 1]
```diff
@@ -113,10 +116,10 @@ class Optimizer(object):
         to sample points, bounds, and type of parameters.
 
     """
-    def __init__(self, dimensions, base_estimator,
+    def __init__(self, dimensions, base_estimator="gp",
                  n_random_starts=None, n_initial_points=10,
                  acq_func="gp_hedge",
-                 acq_optimizer="lbfgs",
+                 acq_optimizer="auto",
```
Comment (betatim) [inline skopt/optimizer/optimizer.py]: doc string update please

File: skopt/optimizer/optimizer.py [hunk 2]
```diff
@@ -175,28 +177,44 @@ def __init__(self, dimensions, base_estimator,
...
         self.n_initial_points_ = n_initial_points
 
-        if (isinstance(base_estimator, (ExtraTreesRegressor,
-                                        RandomForestRegressor,
-                                        GradientBoostingQuantileRegressor)) and
-                not acq_optimizer == "sampling"):
-            raise ValueError(
-                "The tree-based regressor {0} should run with "
-                "acq_optimizer='sampling'".format(type(base_estimator)))
+        tree_ests = (
```
Comment (betatim) [inline skopt/optimizer/optimizer.py]: -> `tree_estimators` or should it be something else because for a GP with categoricals we also use sampling right? `piecewise_constant_estimators` ... too long :-/

File: skopt/optimizer/optimizer.py [hunk 3]
```diff
@@ -175,28 +177,44 @@ def __init__(self, dimensions, base_estimator,
...
+            elif (isinstance(self.base_estimator_, GaussianProcessRegressor)
+                  and self.space.is_categorical):
+                acq_optimizer = "sampling"
+            else:
+                acq_optimizer = "lbfgs"
 
         if acq_optimizer not in ["lbfgs", "sampling"]:
             raise ValueError("Expected acq_optimizer to be 'lbfgs' or "
-                             "'sampling', got {0}".format(acq_optimizer))
+            "'sampling', got {0}".format(acq_optimizer))
```
Comment (betatim) [inline skopt/optimizer/optimizer.py]: formatting mishap?

File: skopt/optimizer/optimizer.py (hunk 2 above)
Comment (betatim) [inline skopt/optimizer/optimizer.py]: Maybe we make a small helper `piecewise_constant(estimator)` that wraps this and the check for categorical GPs? Would also help with the error message on L211 which needs to mention GPs as well

File: skopt/optimizer/optimizer.py [hunk 4]
```diff
@@ -210,7 +228,7 @@ def copy(self, random_state=None):
 
         optimizer = Optimizer(
             dimensions=self.space.dimensions,
-            base_estimator=self.base_estimator,
+            base_estimator=self.base_estimator_,
```
Comment (betatim) [inline skopt/optimizer/optimizer.py]: 👍

File: skopt/utils.py [hunk 5]
```diff
@@ -200,3 +210,39 @@ def func(x):
             best_fun = r.fun
 
     return [v for v in best_x], best_fun
+
+
+def cook_skopt_estimator(base_estimator, n_dims=1, is_cat=False, **kwargs):
```
Comment (betatim) [inline skopt/utils.py]: can we rename it to `make_estimator` or `make_surrogate`? `cook_` makes me think we are in Breaking Bad.

Comment (betatim) [inline skopt/utils.py]: Document me 📖

**Comment by betatim** [comment]
Starting to look good. Will definitely make life easier for ask-and-tell fans.

File: skopt/optimizer/gp.py [hunk 6]
```diff
@@ -223,24 +223,11 @@ def gp_minimize(func, dimensions, base_estimator=None,
...
-
-    base_estimator = GaussianProcessRegressor(
-        kernel=cov_amplitude * other_kernel,
-        normalize_y=True, random_state=rng, alpha=0.0, noise=noise,
-        n_restarts_optimizer=2)
+    base_estimator = cook_skopt_estimator(
+        "GP", is_cat=is_cat, random_state=rng, noise=noise,
+        n_dims=space.transformed_n_dims)
+    if is_cat:
+        acq_optimizer = "sampling"
```
Comment (betatim) [inline skopt/optimizer/gp.py]: can we remove this here and leave it to `Optimizer` to correctly select it based on `acq_optimizer='auto'`?

File: skopt/utils.py (hunk 5 above)
Comment (MechCoder) [inline skopt/utils.py]: Cmon @betatim `cook` sounds cooler.

File: skopt/utils.py [hunk 7]
```diff
@@ -200,3 +210,62 @@ def func(x):
             best_fun = r.fun
 
     return [v for v in best_x], best_fun
+
+
+def cook_estimator(base_estimator, space=None, **kwargs):
```
Comment (MechCoder) [inline skopt/utils.py]: SAY MY NAME!

**Comment by betatim** [comment]
There isn't much "boiler plate" in https://github.com/scikit-optimize/scikit-optimize/blob/master/examples/ask-and-tell.ipynb but should we update it to use this new function so that people learn it exists?

File: skopt/utils.py (hunk 7 above)
Comment (betatim) [inline skopt/utils.py]: Do what? I don't have a clue who you are.

**Comment by MechCoder** [comment]
good idea. I'll do that after you or someone else has made another pass.

File: skopt/utils.py (hunk 7 above)
Comment (betatim) [inline skopt/utils.py]: Joking aside, I still prefer `make_estimator` or even better `make_surrogate`

File: skopt/optimizer/optimizer.py (hunk 2 above)
Comment (betatim) [inline skopt/optimizer/optimizer.py]: `has_gradients()` then or `is_continuous` (in the sense of a continuous function is differentiable)?

File: skopt/utils.py (hunk 7 above)
Comment (glouppe) [inline skopt/utils.py]: I dont mind `cook_estimator` :-) 
(This name also kinda imply that we are cooking some black magic inside to come up with a good default, which is nice.)

File: skopt/utils.py [hunk 8]
```diff
@@ -200,3 +210,62 @@ def func(x):
...
+    * `base_estimator` ["GP", "RF", "ET", "GBRT" or sklearn regressor, default="gp"]:
+        Should inherit from `sklearn.base.RegressorMixin`.
+        In addition the `predict` method, should have an optional `return_std`
+        argument, which returns `std(Y | x)`` along with `E[Y | x]`.
+        If provided base_estimator is in ["GP", "RF", "ET", "GBRT"]
+        then the corresponding estimator used in the minimize functions
+        are used.
+
+    * `space` [Space instance]:
+        Has to be provided is the base_estimator is a gaussian process.
```
Comment (glouppe) [inline skopt/utils.py]: is -> if

File: skopt/optimizer/optimizer.py [hunk 9]
```diff
@@ -44,10 +47,13 @@ class Optimizer(object):
         - an instance of a `Dimension` object (`Real`, `Integer` or
           `Categorical`).
 
-    * `base_estimator` [sklearn regressor]:
+    * `base_estimator` ["GP", "RF", "ET", "GBRT" or sklearn regressor, default="gp"]:
```
Comment (glouppe) [inline skopt/optimizer/optimizer.py]: `default="GP"` (in capital letters, for consistency)

File: skopt/utils.py [hunk 10]
```diff
@@ -200,3 +210,62 @@ def func(x):
...
     return [v for v in best_x], best_fun
+
+
+def cook_estimator(base_estimator, space=None, **kwargs):
+    """
+    Cook a default estimator.
+
+    Parameters
+    ----------
+    * `base_estimator` ["GP", "RF", "ET", "GBRT" or sklearn regressor, default="gp"]:
```
Comment (glouppe) [inline skopt/utils.py]: "gp" -> "GP"

File: skopt/optimizer/optimizer.py [hunk 11]
```diff
@@ -44,10 +47,13 @@ class Optimizer(object):
...
         Should inherit from `sklearn.base.RegressorMixin`.
         In addition the `predict` method, should have an optional `return_std`
         argument, which returns `std(Y | x)`` along with `E[Y | x]`.
+        If provided base_estimator is in ["GP", "RF", "ET", "GBRT"]
+        then the corresponding estimator used in the minimize functions
+        are used.
```
Comment (glouppe) [inline skopt/optimizer/optimizer.py]: that sentence is difficult to read. 

Rather: "[...] then the corresponding estimator is used as surrogate model."?

File: skopt/tests/test_optimizer.py [hunk 12]
```diff
@@ -140,3 +140,19 @@ def test_exhaust_initial_calls():
...
+                        n_random_starts=1)
+    assert "'RF', 'ET' or 'GP'" in str(e.value)
+
+
+@pytest.mark.fast_test
+@pytest.mark.parametrize("base_estimator", ["GP", "RF", "ET"])
+def test_optimizer_base_estimator_string_smoke(base_estimator):
+    opt = Optimizer([(-2.0, 2.0)], base_estimator=base_estimator,
+                    n_random_starts=1)
+    opt.run(func=lambda x: x[0]**2, n_iter=3)
```
Comment (glouppe) [inline skopt/tests/test_optimizer.py]: Can you add augment the test with the lower case strings as well?

Comment (glouppe) [inline skopt/tests/test_optimizer.py]: also add gbrt?

Comment (glouppe) [inline skopt/tests/test_optimizer.py]: or shall we rather directly test `cook_estimator`?

File: skopt/utils.py (hunk 7 above)
Comment (MechCoder) [inline skopt/utils.py]: Sorry to inform you @betatim it seems like you are on the minority on this.

**Comment by MechCoder** [comment]
done

File: skopt/optimizer/optimizer.py [hunk 13]
```diff
@@ -44,10 +48,13 @@ class Optimizer(object):
...
           `Categorical`).
 
-    * `base_estimator` [sklearn regressor]:
+    * `base_estimator` ["GP", "RF", "ET", "GBRT" or sklearn regressor, default="GP"]:
         Should inherit from `sklearn.base.RegressorMixin`.
         In addition the `predict` method, should have an optional `return_std`
         argument, which returns `std(Y | x)`` along with `E[Y | x]`.
+        If provided base_estimator is in ["GP", "RF", "ET", "GBRT"]
+        then the corresponding estimator in the minimize functions is used
+        as a surrogate model.
```
Comment (betatim) [inline skopt/optimizer/optimizer.py]: -> "If base_estimator is one of [GP, ET,...] a default surrogate model of the corresponding type is used. The defaults correspond to what is used in the `*_minimize` functions.

Still not perfect, but I think already an improvement. Need to think a bit more about wording.

File: skopt/utils.py [hunk 14]
```diff
@@ -200,3 +210,72 @@ def func(x):
...
+
+def cook_estimator(base_estimator, space=None, **kwargs):
+    """
+    Cook a default estimator.
+
+    Parameters
+    ----------
+    * `base_estimator` ["GP", "RF", "ET", "GBRT" or sklearn regressor, default="GP"]:
+        Should inherit from `sklearn.base.RegressorMixin`.
+        In addition the `predict` method, should have an optional `return_std`
```
Comment (betatim) [inline skopt/utils.py]: no "," needed after "method"

File: skopt/utils.py [hunk 15]
```diff
@@ -200,3 +210,72 @@ def func(x):
...
     return [v for v in best_x], best_fun
+
+
+def has_gradients(estimator, space):
+    tree_estimators = (
+            ExtraTreesRegressor, RandomForestRegressor,
+            GradientBoostingQuantileRegressor)
+    return (
+        isinstance(estimator, tree_estimators) or
+        (isinstance(estimator, GaussianProcessRegressor) and space.is_categorical)
```
Comment (betatim) [inline skopt/utils.py]: Was just wondering: should we be checking if one of the kernels in the GP is a Hamming kernel to decide if there are gradients available or not? Don't know enough about the various kernels to know if Hamming could have gradients or not. Would also cover weird cases where someone uses a kernel without gradients but doesn't have a categorical space (if such kernels exist).

Tim's mission to reduce the number of function arguments 😆

**Comment by glouppe** [comment]
Hmm Travis is no longer happy. Weird..

**Comment by MechCoder** [comment]
Will merge when tests pass.

File: examples/ask-and-tell.ipynb [hunk 16]
```diff
@@ -139,7 +139,7 @@
    "outputs": [],
    "source": [
     "etr = ExtraTreesRegressor(random_state=2, n_estimators=200)\n",
-    "opt = Optimizer([(-2.0, 2.0)], etr, acq_optimizer=\"sampling\")"
+    "opt = Optimizer([(-2.0, 2.0)], \"ET\", acq_optimizer=\"sampling\")"
```
Comment (glouppe) [inline examples/ask-and-tell.ipynb]: We should have removed the construction of `etr` above.

**Comment by glouppe** [comment]
Nice! great job :)

File: examples/ask-and-tell.ipynb (hunk 16 above)
Comment (rene-rex) [inline examples/ask-and-tell.ipynb]: `etr` is unused now an can be removed, right?

Comment (betatim) [inline examples/ask-and-tell.ipynb]: -> #414
